plugin/clone/src/clone_common.cc

```diff
@@ -0,0 +1,114 @@
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#include "plugin/clone/include/clone_common.h"
+
+#include <string>
+#include <utility>
+
+#include "lex_string.h"
+#include "my_dbug.h"
+#include "my_inttypes.h"
+#include "plugin/clone/include/clone_hton.h"
+#include "sql/handler.h"
+#include "sql/json_dom.h"
+#include "sql/sql_plugin_ref.h"
+#include "storage/perfschema/table_log_status.h"
+
+namespace myclone {
+
+// Call precopy for the participating storage engines that have the
+// precopy method.
+int Ha_clone_common_cbk::precopy(THD *thd, uint task_id) {
+  const auto &all_locators = get_all_locators();
+
+  // Must be called by InnoDB the clone-driving engine
+  assert(all_locators[0].m_hton->db_type == DB_TYPE_INNODB);
+  assert(get_loc_index() == 0);
+  auto *const saved_cbk_hton = get_hton();
+
+  uint index = 0;
+  for (const auto &loc : all_locators) {
+    auto *const precopy_fn = loc.m_hton->clone_interface.clone_precopy;
+    if (precopy_fn) {
+      set_loc_index(index);
+      const auto res =
+          precopy_fn(loc.m_hton, thd, loc.m_loc, loc.m_loc_len, task_id, this);
+      if (res != 0) {
+        set_loc_index(0);
+        set_hton(saved_cbk_hton);
+        return res;
+      }
+    }
+    ++index;
+  }
+  set_loc_index(0);
+  set_hton(saved_cbk_hton);
+
+  return 0;
+}
+
+// Perform the cross-engine synchronization: execute a
+// performance_schema.log_status query, and call set_log_stop for each storage
+// engine with its part of STORAGE_ENGINES column JSON object from that query.
+int Ha_clone_common_cbk::synchronize_engines() {
+  const auto &all_locators = get_all_locators();
+
+  std::unique_ptr<table_log_status> table{static_cast<table_log_status *>(
+      table_log_status::create(&table_log_status::m_share))};
+
+  auto err = table->rnd_init(true);
+  assert(err == 0);
+
+  err = table->rnd_next();
+  if (err != 0) return err;
+
+  auto &log_status_row = table->get_row();
+  const auto &se_positions = log_status_row.w_storage_engines;
+  assert(se_positions.is_dom());
+
+  const auto *const json_dom = se_positions.get_dom();
+  assert(json_dom->json_type() == enum_json_type::J_OBJECT);
+
+  const auto *const json_obj = static_cast<const Json_object *>(json_dom);
+  for (const auto &json_se_pos : *json_obj) {
+    const auto &se_name = json_se_pos.first;
+    const LEX_CSTRING lex_c_se_name{.str = se_name.c_str(),
+                                    .length = se_name.length()};
+    auto *const plugin_ref = ha_resolve_by_name_raw(nullptr, lex_c_se_name);
```

Comment:
I can make this change but I think we are very deep in panic instead of reportable/recoverable error zone here - the server is doomed if it's unable to find its clone-participating storage engines in the middle of clone